LİFE EXPECTANCY (WHO) 

Dünya Sağlık Örgütü (WHO) kapsamındaki Küresel Sağlık Gözlemevi (GHO) veri deposu, sağlık durumu analizi ve tüm ülkeler için ilgili diğer birçok faktörü takip etmektedir. Veri kümeleri, sağlık verileri analizi amacıyla halka açıklanmıştır. 193 ülke için yaşam beklentisi, sağlık faktörleri ile ilgili veri seti aynı DSÖ veri deposu web sitesinden toplanmış ve buna karşılık gelen ekonomik veriler Birleşmiş Milletler web sitesinden toplanmıştır. Sağlıkla ilgili tüm faktör kategorileri arasında sadece daha temsili olan kritik faktörler seçilmiştir. Son 15 yılda, sağlık sektöründe, özellikle gelişmekte olan ülkelerde, son 30 yıla kıyasla insan ölüm oranlarının iyileşmesine yol açan büyük bir gelişme kaydedildiği görülmüştür. Bu nedenle, bu projede daha ileri analizler için 193 ülke için 2000-2015 yılından itibaren verileri dikkate alınmıştır. Tek tek olan veri dosyaları tek bir veri kümesinde birleştirilmiştir. Verilerin ilk görsel muayenesinde bazı eksik değerler görülmüştür. Veri kümeleri DSÖ'den olduğu için belirgin bir hata bulunmamıştır. Eksik veriler R yazılımında Missmap komutu kullanılarak işlenmiştir. Sonuç, eksik verilerin çoğunun popülasyon, Hepatit B ve GSYİH için olduğunu ortaya çıkmıştır. Eksik veriler Vanuatu, Tonga, Togo, Cabo Verde gibi daha az bilinen ülkelerdendir. Bu ülkeler için tüm verileri bulmak zor olduğu için bu ülkeleri nihai model veri kümesinden hariç tutmamıza karar verilmiştir. Son birleştirilmiş dosya (son veri kümesi), 20 tahmin değişkeni anlamına gelen 22 Sütun ve 2938 satırdan oluşmuştur. Tüm tahmin değişkenleri daha sonra birkaç geniş kategoriye ayrılmıştır: Bağışıklık ile ilgili faktörler, Ölüm faktörleri, Ekonomik faktörler ve Sosyal faktörler.

PROBLEM

Nitelikleri kullanarak ortalama ömrü hesaplamaktır. Hedef değişkeni “life expectancy” değişkenidir. Diğer değişkenler ise girdi nitelikleri olmuştur. Ülkelerin isimleri ve gelişme durumları kategorik verilerdir.

DEĞİŞKENLER

Country : ülke

Year: yıl

Status(Developed or developing status): durum

Life expectancy: yaşam beklentisi

Adult mortality: her iki cinsiyetin Yetişkin Ölüm Oranları (1000 kişi başına 15 ila 60 yıl arasında ölme olasılığı)

İnfant deaths: bebek ölümleri (1000 kişi başına düşen ölüm sayısı)

Alcohol: alkol (kişi başına düşen alkol (15+) tüketimi (litre saf alkol cinsinden)

Percentage expenditure: harcama yüzdesi ( kişi başına gayri safi yurtiçi hasıla yüzdesi olarak sağlık harcamaları)

Hepatit B: 1 yaşındakilerde Hepatit B aşı kapsamı (%)

Measles: kızamık (1000 kişi başına rapor edilen vaka sayısı)

BMI: tüm popülasyonun ortalama vücut kitle indeksi

Under-five deaths: beş yaş altı ölümler (1000 kişi başına beş yaş altı ölüm sayısı)

Polio: çocuk felci (çocuk felci aşı kapsamı (%) )

Total expenditure: toplam harcama ( toplam devlet harcamalarının yüzdesi olarak genel devlet sağlık harcamaları (%) ) 

Diphtheria: difteri ( difteri tetanoz toksoid ve boğmaca (DTP3) aşı kapsamı (%) ) 

HIV/AIDS: 1000 canlı doğumda ölüm HIV / AIDS (0-4 yaş)

GDP: GSYH ( kişi başı gayri safi  yurtiçi Hasıla (ABD doları))

Population: nüfus ( ülke nüfusu )

Thinness 1-19 years: 10 ila 19 Yaş arası çocuklar ve ergenler arasında incelik yaygınlığı (%)

Thinness 5-9 years: 5-9 Yaş arası çocuklar arasında incelik yaygınlığı (%)

Income composition of resources: kaynakların gelir kompozisyonu (Kaynakların gelir kompozisyonu açısından İnsani Gelişme Endeksi (0 ile 1 arasında değişen indeks))

Schooling: eğitim ( eğitim yılı)

Kaynak: https://www.kaggle.com/kumarajarshi/life-expectancy-who

In [1]:
import findspark
findspark.init()

In [2]:
from pyspark.sql import SparkSession

from pyspark.sql import SparkSession
spark = SparkSession.builder \
.appName("LinearRegression") \
.master("local[4]") \
.config("spark.driver.memory","2g") \
.config("spark.executor.memory","4g") \
.getOrCreate()

VERİ OKUMA

In [3]:
df= spark.read.format("csv") \
.option("header", True) \
.option("İnferSchema", True) \
.option("sep", ",") \
.load("C:\\Users\\busra\\Desktop\\Egitim\\Apache Spark\LİFE.csv")

In [4]:
df.toPandas().head()

Country  Year      Status Life expectancy  Adult Mortality  \
0  Afghanistan  2015  Developing               65             263   
1  Afghanistan  2014  Developing             59.9             271   
2  Afghanistan  2013  Developing             59.9             268   
3  Afghanistan  2012  Developing             59.5             272   
4  Afghanistan  2011  Developing             59.2             275   

  infant deaths Alcohol percentage expenditure Hepatitis B Measles   ...  \
0            62    0.01            71.27962362          65     1154  ...   
1            64    0.01            73.52358168          62      492  ...   
2            66    0.01            73.21924272          64      430  ...   
3            69    0.01             78.1842153          67     2787  ...   
4            71    0.01            7.097108703          68     3013  ...   

  Polio Total expenditure Diphtheria   HIV/AIDS         GDP Population  \
0     6              8.16          65       0.1   584.25921   33736494   
1    58              8.18          62       0.1  612.696514     327582   
2    62              8.13          64       0.1  631.744976   31731688   
3    67              8.52          67       0.1     669.959    3696958   
4    68              7.87          68       0.1   63.537231    2978599   

   thinness  1-19 years  thinness 5-9 years Income composition of resources  \
0                  17.2                17.3                           0.479   
1                  17.5                17.5                           0.476   
2                  17.7                17.7                            0.47   
3                  17.9                  18                           0.463   
4                  18.2                18.2                           0.454   

  Schooling  
0      10.1  
1        10  
2       9.9  
3       9.8  
4       9.5  

[5 rows x 22 columns]

In [5]:
df.printSchema()

root
 |-- Country: string (nullable = true)
 |-- Year: string (nullable = true)
 |-- Status: string (nullable = true)
 |-- Life expectancy : string (nullable = true)
 |-- Adult Mortality: string (nullable = true)
 |-- infant deaths: string (nullable = true)
 |-- Alcohol: string (nullable = true)
 |-- percentage expenditure: string (nullable = true)
 |-- Hepatitis B: string (nullable = true)
 |-- Measles : string (nullable = true)
 |--  BMI : string (nullable = true)
 |-- under-five deaths : string (nullable = true)
 |-- Polio: string (nullable = true)
 |-- Total expenditure: string (nullable = true)
 |-- Diphtheria : string (nullable = true)
 |--  HIV/AIDS: string (nullable = true)
 |-- GDP: string (nullable = true)
 |-- Population: string (nullable = true)
 |--  thinness  1-19 years: string (nullable = true)
 |--  thinness 5-9 years: string (nullable = true)
 |-- Income composition of resources: string (nullable = true)
 |-- Schooling: string (nullable = true)



In [6]:
df = spark.read.csv("C:\\Users\\busra\\Desktop\\Egitim\\Apache Spark\LİFE.csv", header =True, inferSchema=True)
df.printSchema()

root
 |-- Country: string (nullable = true)
 |-- Year: integer (nullable = true)
 |-- Status: string (nullable = true)
 |-- Life expectancy : double (nullable = true)
 |-- Adult Mortality: integer (nullable = true)
 |-- infant deaths: integer (nullable = true)
 |-- Alcohol: double (nullable = true)
 |-- percentage expenditure: double (nullable = true)
 |-- Hepatitis B: integer (nullable = true)
 |-- Measles : integer (nullable = true)
 |--  BMI : double (nullable = true)
 |-- under-five deaths : integer (nullable = true)
 |-- Polio: integer (nullable = true)
 |-- Total expenditure: double (nullable = true)
 |-- Diphtheria : integer (nullable = true)
 |--  HIV/AIDS: double (nullable = true)
 |-- GDP: double (nullable = true)
 |-- Population: double (nullable = true)
 |--  thinness  1-19 years: double (nullable = true)
 |--  thinness 5-9 years: double (nullable = true)
 |-- Income composition of resources: double (nullable = true)
 |-- Schooling: double (nullable = true)



Veri setine baktığımızda 2 tane kategorik değişken vardır. Diğer değişkenler ise numerik değişkenlerdir. 


In [7]:
new_cols = ["Country", "Year", "Status", "label", "AdultMortality",
      "InfantDeaths", "Alcohol", "PercentageExpenditure", "HepatitisB", "Measles", "BMI", "UnderFiveDeaths",
      "Polio", "TotalExpenditure", "Diphtheria", "HIV_AIDS", "GDP", "Population", "Thinness119", "Thinness59",
      "IncomeCompositionOfResources", "Schooling"]




Değişkenler sırası bozulmadan yeniden isimlendirilmiştir. Hedef değişkeni label olarak değiştirilmiştir.

In [8]:
df2 = df.toDF(*new_cols)

In [9]:
df2.toPandas().head()

Country  Year      Status  label  AdultMortality  InfantDeaths  \
0  Afghanistan  2015  Developing   65.0           263.0            62   
1  Afghanistan  2014  Developing   59.9           271.0            64   
2  Afghanistan  2013  Developing   59.9           268.0            66   
3  Afghanistan  2012  Developing   59.5           272.0            69   
4  Afghanistan  2011  Developing   59.2           275.0            71   

   Alcohol  PercentageExpenditure  HepatitisB  Measles  ...  Polio  \
0     0.01              71.279624        65.0     1154  ...    6.0   
1     0.01              73.523582        62.0      492  ...   58.0   
2     0.01              73.219243        64.0      430  ...   62.0   
3     0.01              78.184215        67.0     2787  ...   67.0   
4     0.01               7.097109        68.0     3013  ...   68.0   

   TotalExpenditure  Diphtheria  HIV_AIDS         GDP  Population  \
0              8.16        65.0       0.1  584.259210  33736494.0   
1              8.18        62.0       0.1  612.696514    327582.0   
2              8.13        64.0       0.1  631.744976  31731688.0   
3              8.52        67.0       0.1  669.959000   3696958.0   
4              7.87        68.0       0.1   63.537231   2978599.0   

   Thinness119  Thinness59  IncomeCompositionOfResources  Schooling  
0         17.2        17.3                         0.479       10.1  
1         17.5        17.5                         0.476       10.0  
2         17.7        17.7                         0.470        9.9  
3         17.9        18.0                         0.463        9.8  
4         18.2        18.2                         0.454        9.5  

[5 rows x 22 columns]

Değişkenler türlerine göre ayrılmıştır. Bunun sebebi  p-value değeri 0.05'ten büyük olan değişkenleri silmemizde bize kolaylık sağlamaktadır.

In [10]:
categorical_cols = ["Country", "Status"]
'''
numerical_cols = ["Year", "AdultMortality",
      "InfantDeaths", "Alcohol", "PercentageExpenditure", "HepatitisB", "Measles", "BMI", "UnderFiveDeaths",
      "Polio", "TotalExpenditure", "Diphtheria", "HIV_AIDS", "GDP", "Population", "Thinness119", "Thinness59",
      "IncomeCompositionOfResources", "Schooling"]
label = ["label"]

'''
numerical_cols = ["Year", "AdultMortality",
      "InfantDeaths", "Alcohol", "BMI", "UnderFiveDeaths",
     "TotalExpenditure", "Diphtheria", "HIV_AIDS",
      "IncomeCompositionOfResources", "Schooling"]
label = ["label"]


In [11]:
df2.describe().toPandas().head()

summary      Country                Year      Status              label  \
0   count         2938                2938        2938               2928   
1    mean         None  2007.5187202178352        None  69.22493169398912   
2  stddev         None   4.613840940258099        None  9.523867487824305   
3     min  Afghanistan                2000   Developed               36.3   
4     max     Zimbabwe                2015  Developing               89.0   

       AdultMortality        InfantDeaths             Alcohol  \
0                2928                2938                2744   
1  164.79644808743168  30.303948264125257  4.6028607871720375   
2   124.2920790034219  117.92650131339906   4.052412658755658   
3                   1                   0                0.01   
4                 723                1800               17.87   

  PercentageExpenditure          HepatitisB  ...               Polio  \
0                  2938                2385  ...                2919   
1     738.2512954533823   80.94046121593291  ...   82.55018842069202   
2     1987.914858016194  25.070015593018063  ...  23.428045949468476   
3                   0.0                   1  ...                   3   
4           19479.91161                  99  ...                  99   

     TotalExpenditure         Diphtheria            HIV_AIDS  \
0                2712               2919                2938   
1  5.9381895280235915  82.32408359027065  1.7421034717494939   
2   2.498319672155633   23.7169120685726   5.077784531086547   
3                0.37                  2                 0.1   
4                17.6                 99                50.6   

                  GDP            Population         Thinness119  \
0                2490                  2286                2904   
1   7483.158469138481  1.2753375120052498E7  4.8397038567493205   
2  14270.169341515959   6.101209650842794E7   4.420194947144322   
3             1.68135                  34.0                 0.1   
4         119172.7418         1.293859294E9                27.7   

          Thinness59 IncomeCompositionOfResources           Schooling  
0               2904                         2771                2775  
1  4.870316804407711           0.6275510645976166  11.992792792792786  
2  4.508882086983007           0.2109035551515931  3.3589197211023554  
3                0.1                          0.0                 0.0  
4               28.6                        0.948                20.7  

[5 rows x 23 columns]

Count kısmına baktığımızda her değişkenin aynı değere sahip olmadığı görülmektedir. Bu durum bize veride boşlukların olduğunu söylemektedir.

In [12]:
df3 = df2.na.drop()

In [13]:
df3.describe().toPandas().head()

summary      Country                Year      Status              label  \
0   count         1649                1649        1649               1649   
1    mean         None  2007.8405093996362        None  69.30230442692543   
2  stddev         None   4.087710534059055        None  8.796834135238601   
3     min  Afghanistan                2000   Developed               44.0   
4     max     Zimbabwe                2015  Developing               89.0   

       AdultMortality        InfantDeaths            Alcohol  \
0                1649                1649               1649   
1   168.2152819890843   32.55306246209824  4.533195876288676   
2  125.31041693156172  120.84719049638885  4.029188999208425   
3                   1                   0               0.01   
4                 723                1600              17.87   

  PercentageExpenditure         HepatitisB  ...               Polio  \
0                  1649               1649  ...                1649   
1      698.973558049698  79.21770770163735  ...   83.56458459672528   
2    1759.2293363477393  25.60466388103202  ...  22.450557485191844   
3                   0.0                  2  ...                   3   
4            18961.3486                 99  ...                  99   

     TotalExpenditure         Diphtheria            HIV_AIDS  \
0                1649               1649                1649   
1   5.955924802910845    84.155245603396  1.9838690115221154   
2  2.2993854200363937  21.57919284379662   6.032359689186714   
3                0.74                  2                 0.1   
4               14.39                 99                50.6   

                  GDP            Population        Thinness119  \
0                1649                  1649               1649   
1   5566.031886817592  1.4653625889484541E7  4.850636749545183   
2  11475.900117274561    7.04603934030564E7  4.599228369708468   
3             1.68135                  34.0                0.1   
4         119172.7418         1.293859294E9               27.2   

          Thinness59 IncomeCompositionOfResources           Schooling  
0               1649                         1649                1649  
1  4.907762280169795           0.6315512431776829  12.119890842935117  
2  4.653756745508691          0.18308872643487184   2.795387538952672  
3                0.1                          0.0                 4.2  
4               28.2                        0.936                20.7  

[5 rows x 23 columns]

Count her değişkende de eşit olmuştur. Değişkenlerin özetleyici istatistikleri de gelmiştir. Burada yapılan işlem en kolay veri temizleme işlemidir. İsterseniz ayrıntılı bir şekilde veriyi inceleyip en az miktarda veri setini azaltabilirsiniz.

In [14]:
from pyspark.ml import Pipeline, PipelineModel
from pyspark.ml.regression import LinearRegression, LinearRegressionModel

In [15]:
from pyspark.sql import SparkSession
from pyspark.ml.feature import StringIndexer,VectorIndexer,OneHotEncoder,VectorAssembler

Veri ön hazırlığı için gerekli kütüphaneler indirilmiştir. Country değişkeni burada analize alınmamıştır.

STRİNG INDEXER

In [16]:
status_string_indexer=StringIndexer(inputCol="Status",outputCol="StatusIndex")

ONE HOT ENCODER

In [17]:
encoder=OneHotEncoder(inputCol="StatusIndex",outputCol="StatusEncoded")

In [18]:
'''
usedfeature=["Year", "AdultMortality",
      "InfantDeaths", "Alcohol", "PercentageExpenditure", "HepatitisB", "Measles", "BMI", "UnderFiveDeaths",
      "Polio", "TotalExpenditure", "Diphtheria", "HIV_AIDS", "GDP", "Population", "Thinness119", "Thinness59",
      "IncomeCompositionOfResources", "Schooling","StatusEncoded"]
'''
usedfeature=["Year", "AdultMortality",
      "InfantDeaths", "Alcohol", "BMI", "UnderFiveDeaths",
     "TotalExpenditure", "Diphtheria", "HIV_AIDS",
      "IncomeCompositionOfResources", "Schooling","StatusEncoded"]


VECTOR ASSEMBLER

In [19]:
vector_assembler=VectorAssembler(inputCols=usedfeature,outputCol="features")

LİNEER MODEL

In [20]:
linear_regression_object = LinearRegression().setFeaturesCol("features").setLabelCol("label")

PİPELİNE

In [21]:
pipeline_object = Pipeline(stages = [status_string_indexer,encoder,vector_assembler,linear_regression_object])

VERİ SETİNİ AYIRMA İŞLEMİ

In [22]:
train_df, test_df = df3.randomSplit([0.8, 0.2], seed=142 )

Veri seti train ve test olarak ikiye ayrılmıştır.

In [23]:
train_df.cache()
test_df.cache()

DataFrame[Country: string, Year: int, Status: string, label: double, AdultMortality: int, InfantDeaths: int, Alcohol: double, PercentageExpenditure: double, HepatitisB: int, Measles: int, BMI: double, UnderFiveDeaths: int, Polio: int, TotalExpenditure: double, Diphtheria: int, HIV_AIDS: double, GDP: double, Population: double, Thinness119: double, Thinness59: double, IncomeCompositionOfResources: double, Schooling: double]

Burada veriler belleğe alınmıştır. Bu sayede veri diskte okunur ve zaman kaybı yaşanmaz.

In [24]:
model = pipeline_object.fit(train_df)

Model kurulum işlemi gerçekleşmiştir.

In [25]:
model.transform(test_df).select("label","prediction").toPandas().head()

label  prediction
0   56.7   58.978295
1   58.1   61.942204
2   58.8   62.709066
3   59.9   63.497702
4   75.3   74.281806

Label gerçek yaş değişkenidir. Prediction ise tahmin değerleridir. En ideal modeli seçmek için model tuning işlemi yapılmalıdır. Modele katkısı olmayan değişkenler çıkarılmalıdır.

In [26]:
lr_model = model.stages[-1]

In [27]:
lr_model.summary.pValues

[1.066483899037074e-05,
 0.0,
 3.7978065137167505e-11,
 0.009445434020568078,
 5.474766067159464e-07,
 1.6265655489178243e-11,
 0.0012039774175587326,
 0.0002756966957075946,
 0.0,
 0.0,
 0.0,
 0.000177030389040711,
 5.841720929211647e-08]

In [28]:
zipped = zip(lr_model.summary.pValues, numerical_cols+["Status","Sabit"])

In [29]:
sorted_zip = sorted(zipped)

In [30]:
for i in sorted_zip:
    print(i)

(0.0, 'AdultMortality')
(0.0, 'HIV_AIDS')
(0.0, 'IncomeCompositionOfResources')
(0.0, 'Schooling')
(1.6265655489178243e-11, 'UnderFiveDeaths')
(3.7978065137167505e-11, 'InfantDeaths')
(5.841720929211647e-08, 'Sabit')
(5.474766067159464e-07, 'BMI')
(1.066483899037074e-05, 'Year')
(0.000177030389040711, 'Status')
(0.0002756966957075946, 'Diphtheria')
(0.0012039774175587326, 'TotalExpenditure')
(0.009445434020568078, 'Alcohol')


Burada değişkenlerin p-value değerleri küçükten büyüye sıralanmıştır. P-value değeri en büyük olanlar sırayla veri setinden çıkarılmıştır. Son hali yukarıda görüldüğü gibidir.

In [31]:
lr_model.summary.r2

0.8325712025036596

Değişkenler varken çıkan r2 değeri =0.8315062814076117

Değişkenler silindikten sonra çıkan r2 değeri=0.8325712025036596

R2 değerimiz neredeyse hiç azalmadı. R2‘ın yüksek olması regresyon model uyumunun iyi olduğunu gösterir.

MODEL TUNİNG

In [32]:
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.tuning import CrossValidator, ParamGridBuilder

ParamGrid

In [33]:
param_grid = ParamGridBuilder() \
.addGrid(linear_regression_object.aggregationDepth, [2,5]) \
.addGrid(linear_regression_object.elasticNetParam, [0.0,0.02]) \
.addGrid(linear_regression_object.epsilon, [1.35, 1.55]) \
.addGrid(linear_regression_object.maxIter, [10, 20]) \
.addGrid(linear_regression_object.regParam, [0.00, 0.01, 0.05]) \
.addGrid(linear_regression_object.solver, ["auto", "normal", "l-bfgs"]) \
.addGrid(linear_regression_object.tol, [1.0E-6, 1.0E-4]) \
.build()

Cross-Validation

In [34]:
cv = CrossValidator() \
.setEstimator(pipeline_object) \
.setEvaluator(RegressionEvaluator()) \
.setEstimatorParamMaps(param_grid) \
.setNumFolds(5) \
.setParallelism(2)

In [35]:
cv_model = cv.fit(train_df)

In [36]:
cv_model.transform(test_df).select("label","prediction").toPandas().head()

label  prediction
0   56.7   58.976651
1   58.1   61.941111
2   58.8   62.708207
3   59.9   63.497391
4   75.3   74.282824

En İyi Model

In [37]:
best_model = cv_model.bestModel

In [38]:
type(best_model)

pyspark.ml.pipeline.PipelineModel

In [39]:
lr_model = best_model.stages[-1]

In [40]:
lr_model.coefficients

DenseVector([-0.1138, -0.0174, 0.0826, -0.0967, 0.0321, -0.063, 0.145, 0.0179, -0.4454, 9.9945, 1.0016, -1.3823])

Değişkenlerin katsayı değerleri yukarıda görüldüğü gibidir. P-value değerlerine bakıldığında ise modele katkıları anlamlıdır.

In [41]:
lr_model.intercept

281.2807988208168

En iyi modelin sabit değeri 281.280 çıkmıştır.

In [42]:
lr_model.summary.r2

0.8325711772242368

Değişkenleri kullanarak ortalama ömrün %83.25'ini tahmin edebiliyoruz.

In [43]:
lr_model.summary.rootMeanSquaredError

3.603785488257329

Modelin hata kare ortalaması 3.603 çıkmıştır.

In [44]:
trainingSummary = lr_model.summary
print("RMSE: %f" % trainingSummary.rootMeanSquaredError)
print("r2: %f" % trainingSummary.r2)

RMSE: 3.603785
r2: 0.832571


RMSE, modele göre öngörülen değerler ile gerçek değerler arasındaki farkları ölçer. Böyle bir karşılaştırmadan sonra, RMSE'miz oldukça iyi görünüyor.